### ISSUE TITLE
Enabling DNF Count Me support in Fedora CoreOS

### ISSUE THREAD

**Issue by travier** [coreos/fedora-coreos-tracker#717]
**Describe the enhancement**

rpm-ostree [[1](https://github.com/coreos/rpm-ostree/pull/2372),[2](https://github.com/coreos/rpm-ostree/issues/2251)] now supports the [DNF Count Me](https://fedoraproject.org/wiki/Changes/DNF_Better_Counting) feature to help Fedora anonymously count the number of users of a given Fedora release. This issue os about enabling this by default for Fedora CoreOS. Fedora Silverblue and Fedora IoT will have to discuss and implement this independently.

This is split from https://github.com/coreos/fedora-coreos-tracker/issues/86 as this is only about Count Me support and not about more general hardware and software statistics collection.

**Additional information**

The suggested plan is as follow:

  - Discuss this is in a FCOS community meeting
  - Post announcements about the change to Twitter, mailing list, meeting notes etc.
  - Update the docs to let people disable that beforehand if they don't want to be counted
  - Wait for 2 weeks
  - Enable it for testing for the next release
  - Post announcements again once the stable release with it enabled gets pushed out

**Comment by cgwalters** [comment]
One thing we could try to do is "enable only on new images".  (But the more we do this the more it makes testing a mess etc.)

**Comment by travier** [comment]
One of the discussed option was to make the service also check the config from https://github.com/coreos/fedora-coreos-pinger to make the transition from the existing options there.

**Comment by travier** [comment]
We could also work on libdnf to allow rpm-ostree to disable the built-in count me support to avoid duplication / leaks when overlaying packages and thus making disabling the timer the only logic needed to disable count me on FCOS.

**Comment by travier** [comment]
@dustymabe also suggested to write a Fedora Magazine article to explain the change.

**Comment by travier** [comment]
The easiest way I could find to make libdnf not send countme request would be to not download packages as root, which would be a good thing for other reasons: https://github.com/rpm-software-management/libdnf/blob/dnf-4-master/libdnf/repo/Repo.cpp#L1059

**Comment by tpopela** [comment]
@travier thank you for working on this!

> This issue is about enabling this by default for Fedora CoreOS. Fedora Silverblue and Fedora IoT will have to discuss and implement this independently.

Can you point me at where to enable this for Silverblue? Will it be somewhere in https://pagure.io/workstation-ostree-config ?

**Comment by travier** [comment]
> Can you point me at where to enable this for Silverblue? Will it be somewhere in https://pagure.io/workstation-ostree-config ?

So to enable this we need to enable the `rpm-ostree-countme.timer` somehow. The question is how, and for which systems (existing ones vs new ones). We have several options:

  * Enable it directly in the RPM: Enabled for everybody who consumes rpm-ostree. Will require coordination and will have to be special cased for systems that do not want that.
  * Enable it in system presets: Only apply on firstboot if I'm not mistaken.
  * Something else?

**Comment by tpopela** [comment]
Thank you @travier for all the information! We've talked about this on our Silverblue meeting and we would like to follow the same way as it's done on Fedora Workstation - that means enabled by default and users can opt out (currently using the CLI, but we do plan to incorporate some switch into GNOME Initial Setup to easily opt-out). I'm curious whether we could use fedora-release-silverblue package and enable the systemd presets there.

**Comment by travier** [comment]
Another option is to create a `countme` RPM sub-package with just those units with a default enable. This is easy to include in FCOS/IoT/Silverblue and easy to keep out of Workstation.

**Comment by tpopela** [comment]
Good idea @travier ! And users could easily opt out by removing the package..

**Comment by travier** [comment]
Indeed! But that would not enable a "before the change" opt out path. I'm afraid we will still need explicit masking recommandations for users that want to opt-out.

**Comment by travier** [comment]
There is another option: Adding a `Wants=rpm-ostree-countme.timer` config snippet for `rpm-ostreed` unit to fedora-releases packages. This is per variant and will be only enabled for rpm-ostree ones.

**Comment by travier** [comment]
Working on that one. Thanks @cgwalters for the suggestion.

**Comment by travier** [comment]
https://src.fedoraproject.org/rpms/fedora-release/pull-request/173 @cgwalters @tpopela

**Comment by travier** [comment]
https://src.fedoraproject.org/rpms/rpm-ostree/pull-request/43

**Comment by travier** [comment]
Previous discussion: https://meetbot-raw.fedoraproject.org/teams/fedora_coreos_meeting/fedora_coreos_meeting.2021-01-20-16.30.html

**Comment by jlebon** [comment]
This was discussed during today's community meeting:

> #agreed we will enable countme support by default in upgrades and new installs during the f34 cycle. we will announce this on the mailing lists, twitter account, forum, and fedmag at least 3 months before with details about how to opt-out ahead of the rebase.

@bgilbert mentioned we should also include it in the f34 release notes.

**Comment by travier** [comment]
Draft document for Fedora Magazine article, mailing list posts, etc: https://hackmd.io/yGOttCrHQeO9jBrRujJxHA

Review for wording, etc. welcomed! 🙂

**Comment by travier** [comment]
https://discussion.fedoraproject.org/t/article-proposal-getting-better-at-counting-rpm-ostree-based-systems/28681

**Comment by travier** [comment]
Probably the last chance for review / edits before the article gets published.

**Comment by travier** [comment]
The Fedora Magazine will soon be published. Setting the not before date for count me being enabled in FCOS at "not before August 1st".

**Comment by travier** [comment]
https://fedoramagazine.org/getting-better-at-counting-rpm-ostree-based-systems/

**Comment by travier** [comment]
https://hackmd.io/yGOttCrHQeO9jBrRujJxHA > Draft for coreos-status ML

**Comment by travier** [comment]
https://github.com/coreos/fedora-coreos-docs/pull/289

**Comment by dustymabe** [comment]
Mailing list post: https://lists.fedoraproject.org/archives/list/coreos-status@lists.fedoraproject.org/message/OC5XVAEXAI3MWISBDPCSTAF2OW67XW77/

**Comment by SvenMW** [comment]
That this is OPT OUT instead of OPT IN might be a GDPR violation.

the counting mechansim is pooly explained here:

https://fedoraproject.org/wiki/Changes/DNF_Better_Counting

but at first it says:
> "For this reason, we don’t want to use any identifier like /etc/machine-id which may be used for other purposes — or in fact any UUID at all"

but later it says:

> "Older versions will not have the UUID counting enabled"

it also lists the following information, but this does not seem to be a complete list, where can I see the complete list of data transmitted without grepping source code?

> "Add VARIANT_ID (see Changes/Label Our Variants) to string reported to when metadata is requested from fedora update servers
Current requests include machine architecture and Fedora OS version as part of the path; we may want to also put those in a standard format for easy processing (implementation detail)"

This information might, in fact be PII (personally identifiable information) according to gdpr, because the combination of architecture variant id, maybe IP address and fedora os version might very well be unique for a given network.

**Comment by dustymabe** [comment]
The change has been reviewed by Legal and is something that the fedora project has taken on as a whole. In general we really _aren't_ interested in tracking people, or making it possible to track people. The [dnf  documentation](https://dnf.readthedocs.io/en/latest/conf_ref.html) might have more details you'd be interested in (search for `countme`).

If you have more questions we do have a [Fedora Legal mailing list](https://lists.fedoraproject.org/archives/list/legal@lists.fedoraproject.org/).

**Comment by dustymabe** [comment]
Looks like we can proceed with this around the August 14th timeline. Should we start staging things in for our upcoming next few releases?

**Comment by travier** [comment]
Yes, I'm planning on enabling that for the next testing/next releases.

**Comment by travier** [comment]
See also https://github.com/coreos/rpm-ostree/issues/3033

**Comment by travier** [comment]
@cgwalters I had initially dismissed using systemd presets to enable the timer due to an incorrect assumption that they would only be applied on first boot on rpm-ostree base systems. Now that we a have an `ostree-counting` variant specific subpackage in `fedora-release`, should we switch the indirect enabling via a require to a preset as this would make this conceptually simpler?

**Comment by cgwalters** [comment]
Hmm.  I thought the preset bits were the same between traditional and rpm-ostree.  (Though, there's also some CoreOS-vs-Anaconda bits going on here too, and actually in fedora-coreos-config we explicitly `preset-all` in the base ostree commit to undo random RPM scripts doing `systemctl enable`)

Now, the [systemd rpm macros](https://github.com/systemd/systemd/blob/f6e40037a0c6f027fd30e141092dd4982cc56050/src/rpm/macros.systemd.in#L51) basically boils down to running `systemctl --no-reload preset` *only* on the first time a package is installed.

The intention I think is to enable at that "point in time" (pairing of package + preset state), which creates hysteresis.  In rpm-ostree we generally avoid hysteresis, so I think actually what will happen here with rpm-ostree is regardless of whether a package is layered or not, we actually will e.g. enable the service if the preset starts enabling it, and correspondingly disable if the preset no longer enables it.

**Comment by cgwalters** [comment]
A simple way to think about all this is that the semantics are pretty similar to if one does a container build `podman/docker build` from a `Dockerfile` that does `yum -y install foo` and used that for updates.  Each update appears as a fresh install, so the preset logic is executed each time.

**Comment by travier** [comment]
We have our own presets for FCOS and I could create an ostree-countme preset for other rpm-ostree based editions.

**Comment by cgwalters** [comment]
Isn't the DNF equivalent enabled globally in all Fedora editions?  Why not just enable the rpm-ostree equivalent globally too in `90-default.preset` or so?

**Comment by travier** [comment]
Preset PR in https://github.com/coreos/fedora-coreos-config/pull/1167

**Comment by cgwalters** [comment]
> and we should not enable the rpm-ostree count me timer on Workstation but only Silverblue.

True, but we already have `ConditionPathExists=/run/ostree-booted` in the timer unit, so having it enabled won't do anything either.

**Comment by jlebon** [comment]
We also already have other ostree units enabled there:
- https://src.fedoraproject.org/rpms/fedora-release/blob/rawhide/f/90-default.preset#_250
- https://src.fedoraproject.org/rpms/fedora-release/blob/rawhide/f/90-default.preset#_306

Having the timer preset sit alongside `enable dnf-makecache.timer` makes sense to me. (For FCOS, seems fine to fast-track it if we really want as in https://github.com/coreos/fedora-coreos-config/pull/1167/ until it gets there.)

**Comment by travier** [comment]
Why not do the reverse and split all ostree specific enables into another preset file now that we essentially have an "ostree" release sub-package?

**Comment by cgwalters** [comment]
If all rpm-ostree using editions want countme enabled by default (and I believe they do because it's a Fedora default), then what's the value in splitting it into a separate preset file?

I'm not opposed, I just don't see the value in doing so.  A minor negative is that it perpetuates the mindset that ostree is somehow "secondary".

Remember that these presets are just *defaults* - preset files have a nice override mechanism, so anyone e.g. wanting to make a custom rpm-ostree system using Fedora content can easily have their own preset file which disables the service by default without needing it split into a separate package that they remove.

For example in FCOS we ship [this preset file](https://github.com/coreos/fedora-coreos-config/blob/testing-devel/overlay.d/05core/usr/lib/systemd/system-preset/40-coreos-systemd.preset) which overrides some of the base Fedora defaults for systemd.  Someone deriving from FCOS could easily flip them back on in e.g. `20-my-custom-fcos-systemd.preset` too.

**Comment by travier** [comment]
OK, I would have preferred to not enable a unit on systems where it should and would not run but enabling this globaly will indeed make things simpler. I'll update the `fedora-release` package and the temporary, fast-track FCOS PR.

**Comment by cgwalters** [comment]
Remember too, presets only take effect if the software is installed (e.g. for systems where `rpm-ostree` isn't installed at all, the service obviously won't actually run).   I think most traditional systems don't have it installed.

*And* we still have the conditionals.

There are a lot of other services in this case, e.g. we have presets enabling `lvm2` stuff that are just no-ops if lvm isn't installed.

I would say again though, I am not *opposed* to splitting the preset file if you feel strongly it should be done.  But if we're going that route, I'd question why not try to e.g. also split up the preset files for "desktop stuff" vs "baseos stuff" etc.  I mean, reading `90-default.preset` it's just a huge grab bag of random stuff.

**Comment by travier** [comment]
https://src.fedoraproject.org/rpms/fedora-release/pull-request/199

**Comment by travier** [comment]
This has been enabled in this week testing release: 34.20210821.2.0

**Comment by travier** [comment]
@dustymabe I've updated the notes at https://hackmd.io/yGOttCrHQeO9jBrRujJxHA as draft for the list announcement

**Comment by dustymabe** [comment]
The fix for this went into stable stream release `34.20210821.3.0`.
